src/adapters/scene_category.js

```diff
@@ -1,21 +1,66 @@
-import { Marker } from 'mapbox-gl--ENV';
 import constants from '../../config/constants.yml';
-import { createIcon } from '../adapters/icon_manager';
 import Telemetry from 'src/libs/telemetry';
 import { toUrl } from 'src/libs/pois';
 import { fire, listen } from 'src/libs/customEvents';
+import { poisToGeoJSON, emptyFeatureCollection } from 'src/libs/geojson';
+import { filteredPoisStyle, hoveredPoiStyle } from 'src/adapters/pois_styles';
+import { createDefaultMapIcon } from 'src/adapters/icon_manager';
 
-function getMarkerId(poi) {
-  return `marker_${poi.id}`;
-}
+const DYNAMIC_POIS_LAYER = 'poi-filtered';
+const ACTIVE_POIS_LAYER = 'poi-active';
 
 export default class SceneCategory {
   constructor(map) {
     this.map = map;
-    this.markers = [];
+
+    createDefaultMapIcon().then(imageData => {
+      this.map.addImage('pin_with_dot', imageData);
+    });
+
+    this.map.addSource(DYNAMIC_POIS_LAYER, {
+      type: 'geojson',
+      data: emptyFeatureCollection,
+      promoteId: 'id', // tells MapBox-GL to use this property as internal feature identifier
+    });
+    this.map.addLayer({
+      ...filteredPoisStyle,
+      source: DYNAMIC_POIS_LAYER,
+      id: DYNAMIC_POIS_LAYER,
+    });
+    this.map.on('click', DYNAMIC_POIS_LAYER, e => {
+      e.cancelMapClick = true; // Hack as MapBox events don't have stopPropagation
+      const poi = this.getPointedPoi(e);
+      this.selectPoi({
+        poi,
+        pois: this.pois,
+        poiFilters: this.poiFilters,
+      });
+    });
+    this.map.on('mouseenter', DYNAMIC_POIS_LAYER, e => {
+      this.map.getCanvas().style.cursor = 'pointer';
+      const poi = this.getPointedPoi(e);
+      this.highlightPoiMarker(poi, true);
+      fire('open_popup', this.getPointedPoi(e), e.originalEvent);
+    });
+    this.map.on('mouseleave', DYNAMIC_POIS_LAYER, e => {
+      this.map.getCanvas().style.cursor = '';
+      const poi = this.getPointedPoi(e);
+      this.highlightPoiMarker(poi, false);
+      fire('close_popup');
+    });
+
+    this.map.addSource(ACTIVE_POIS_LAYER, {
+      type: 'geojson',
+      data: emptyFeatureCollection,
+      promoteId: 'id',
+    });
+    this.map.addLayer({
+      ...hoveredPoiStyle,
+      source: ACTIVE_POIS_LAYER,
+      id: ACTIVE_POIS_LAYER,
+    });
```

Comment:
Is it really necessary to define a separate source and layer to represent the active POI ?
This is a use-case where the "feature state" mechanism provided by mapbox-gl is supposed to be appropriate.
https://docs.mapbox.com/help/tutorials/create-interactive-hover-effects-with-mapbox-gl-js/

And the performance would probably be better too, as the source data would be updated less often.

Reply:
This is what I explained previously, this is indeed something that looks like a use-case for feature state. So that's what I tried happily at first… only to re-discover that `icon-size` and `icon-image` are "layout" properties, that don't support being set with `feature-state` (probably as it requires too much redraws).
That's another of the points where the Mapbox-GL doc is hard to parse and not explicit enough.
This is the closest official inspiration I used: https://docs.mapbox.com/android/maps/examples/icon-size-change-on-click/ (the same case but on the Android SDK :/)